sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/executeImmediate.scala

```diff
@@ -17,190 +17,137 @@
-        } else if (posNodes.nonEmpty && namedNodes.nonEmpty) {
-          throw QueryCompilationErrors.invalidQueryMixedQueryParameters()
-        } else {
-          if (posNodes.nonEmpty) {
-            PosParameterizedQuery(plan, expressions)
-          } else {
-            val aliases = expressions.collect {
-              case e: Alias => e
-              case u: VariableReference => Alias(u, u.identifier.name())()
-            }
-
-            if (aliases.size != expressions.size) {
-              val nonAliases = expressions.filter(attr =>
-                !attr.isInstanceOf[Alias] && !attr.isInstanceOf[VariableReference])
+      case e @ ExecuteImmediateQuery(queryParam, args, targetVariables) =>
+        // Check if all expressions are resolved (they should be resolved by ResolveReferences now)
+        val queryParamResolved = queryParam.resolved
+        val allArgsResolved = args.forall(_.resolved)
+        val targetVariablesResolved = targetVariables.forall {
+          case _: UnresolvedAttribute => false // Unresolved attributes are not resolved
+          case alias: Alias => alias.child.resolved // For aliases, check if child is resolved
+          case _: VariableReference => true // VariableReference is already resolved
+          case expr => expr.resolved // For other expressions, use standard resolved check
+        }
 
-              throw QueryCompilationErrors.invalidQueryAllParametersMustBeNamed(nonAliases)
+        // Validate that USING clause expressions don't contain unsupported constructs
+        validateUsingClauseExpressions(args)
+
+        // Validate that query parameter is foldable (constant expression)
+        validateQueryParameter(queryParam)
+
+        if (queryParamResolved && allArgsResolved && targetVariablesResolved) {
+          // All resolved - transform based on whether we have target variables
+          if (targetVariables.nonEmpty) {
+            // EXECUTE IMMEDIATE ... INTO should generate SetVariable plan
+            // SetVariable expects UnresolvedAttribute objects that ResolveSetVariable will resolve
+            val finalTargetVars = targetVariables.map {
+              case attr: UnresolvedAttribute =>
+                // Keep as UnresolvedAttribute for ResolveSetVariable to handle
+                attr
+              case alias: Alias =>
+                // Extract the UnresolvedAttribute from the alias
+                alias.child match {
+                  case attr: UnresolvedAttribute =>
+                    attr
+                  case varRef: VariableReference =>
+                    // Convert back to UnresolvedAttribute for ResolveSetVariable
+                    UnresolvedAttribute(varRef.originalNameParts)
+                  case _ =>
+                    throw QueryCompilationErrors.unsupportedParameterExpression(alias.child)
+                }
+              case varRef: VariableReference =>
+                // Convert back to UnresolvedAttribute for ResolveSetVariable
```

Comment:
I don't get it. Why can't we construct resolved `SetVariable` directly with `VariableReference`? We can change `VariableReference#canFold` to false before putting it into `SetVariable`.

Reply:
Yes, that was convoluted.